userspace/libsinsp/metrics_collector.h

```diff
@@ -0,0 +1,246 @@
+	SINSP_STATS_V2_N_DROPS_FULL_THREADTABLE, ///< Number of drops due to full threadtable, unit: count.
+	SINSP_STATS_V2_N_MISSING_CONTAINER_IMAGES, ///<  Number of cached containers (cgroups) without container info such as image, hijacked sinsp_container_manager::remove_inactive_containers() -> every flush snapshot update, unit: count.
+	SINSP_STATS_V2_N_CONTAINERS, ///<  Number of containers (cgroups) currently cached by sinsp_container_manager, hijacked sinsp_container_manager::remove_inactive_containers() -> every flush snapshot update, unit: count.
+	SINSP_MAX_STATS_V2
+};
+
+#ifdef __linux__
+
+namespace libsinsp::metrics {
+
+class metrics_collector
+{
+public:
+	metrics_collector(sinsp* inspector, uint32_t flags, bool convert_memory_to_mb);
+
+	/*!
+	\brief Method to fill up m_metrics_buffer with metrics; refreshes m_metrics with up-to-date metrics on each call
+	*/
+	void snapshot();
+
+	/*!
+	\brief Method to get a const reference to m_metrics vector
+	*/
+	const std::vector<metrics_v2>& get_metrics() const;
+
+	/*!
+	\brief Method to convert a metric to the text-based Prometheus exposition format.
+	 * 	
+	 * Reference: https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+	 * Note: The design idea is to expose Prometheus metrics by piping text-based formats to new line-delimited fields
+	 * exposed at /metrics in Falco's existing HTTP webserver (w/ optional mTLS support), eliminating the need for implementing 
+	 * a complete Prometheus client.
+	 * 
+	 * We exclusively support counter and gauge Prometheus metric types, covering metrics from kernel driver tracepoints
+	 * to linsinsp and client metrics. Introducing a registry seems excessive, especially given the dynamic nature of the final
+	 * metric string names, such as variations in tracepoints across architectures.
+	 * Considering the simplistic use case, adding another dependency to the project does not seem justified. Furthermore, for C++ 
+	 * (compared to Go for example), there appear to be fewer formal client library projects available. Plus, we need to think 
+	 * about stability and long-term support before adding any new dependency.
+	 * 
+	 * The final fully qualified Prometheus metric name partially follows https://prometheus.io/docs/practices/naming/
+	 * Prepend namespace and subsystem with "_" delimiter to create a fully qualified metric name according to 
+	 * https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Opts + append unit with "_" delimiter
+	 * We do not follow the concept of base_units, but guarantee no units are mixed per unique `prom_metric_name_fully_qualified`
+	 * 
+	 * Example:
+	 * 
+	 * # HELP testns_falco_n_threads_total https://falco.org/docs/metrics/
+	 * # TYPE testns_falco_n_threads_total gauge
+	 * testns_falco_n_threads_total{raw_name="n_threads",example_key1="example1",example_key2="example2"} 12 1707281978248705000
```

Comment:
So this made me wonder the following about the timestamp value:
* what's the expected format?
* should it be included?

The first one is easy enough to answer [from the docs](https://prometheus.io/docs/instrumenting/exposition_formats/#comments-help-text-and-type-information):
> The timestamp is an int64 (milliseconds since epoch, i.e. 1970-01-01 00:00:00 UTC, excluding leap seconds), represented as required by Go's [ParseInt()](https://golang.org/pkg/strconv/#ParseInt) function.

So it looks like we'd need to change that value to be in milliseconds instead of nanoseconds.

The second question is a bit trickier. The field is optional. When Prometheus will scrape your metrics, if the timestamp is there, the data will be inserted at that timestamp. But if it isn't, the data is assumed to be valid at the exact time Prometheus scrapes it. I don't know which is best here, but it's worth considering which you'd prefer. I think it'd be best to only fill it in if the data you're exposing is not in sync with the current state of the system, if that makes sense?

Reply:
Reading this, I think we can just remove it. The client who would call this method (Falco repo) has a timer and would override and refresh the data under `/metrics` every pre-defined ticker period. And FYI on the Falco side we are already compliant with Prometheus time duration notations for consistency and clarity in adoption, see [here](https://github.com/falcosecurity/falco/blob/b091522398903175103f99251552dea84a3635ef/userspace/engine/falco_utils.cpp#L57) (thanks to @jasondellaluce who called that out).